wp-includes/rest-api/auth/class-wp-rest-token.php

```diff
@@ -0,0 +1,703 @@
+							'type'        => 'object',
+							'readonly'    => true,
+							'properties'  => array(
+								'id'         => array(
+									'description' => esc_html__( 'The ID of the user.', 'jwt-auth' ),
+									'type'        => 'integer',
+									'readonly'    => true,
+								),
+								'type'       => array(
+									'description' => esc_html__( 'The type of user.', 'jwt-auth' ),
+									'type'        => 'string',
+									'readonly'    => true,
+								),
+								'user_login' => array(
+									'description' => esc_html__( 'The username of the user.', 'jwt-auth' ),
+									'type'        => 'string',
+									'readonly'    => true,
+								),
+								'user_email' => array(
+									'description' => esc_html__( 'The email address of the user.', 'jwt-auth' ),
+									'type'        => 'string',
+									'readonly'    => true,
+								),
+								'api_key'    => array(
+									'description' => esc_html__( 'The API key of the user.', 'jwt-auth' ),
+									'type'        => 'string',
+									'readonly'    => true,
+								),
+							),
+						),
+					),
+				),
+			),
+		);
+
+		/**
+		 * Filters the REST endpoint schema.
+		 *
+		 * @param string $schema The endpoint schema.
+		 */
+		return apply_filters( 'rest_authentication_token_schema', $schema );
+	}
+
+	/**
+	 * Authenticate and determine if the REST request has authentication errors.
+	 *
+	 * @filter rest_authentication_errors
+	 *
+	 * @since 0.1
+	 *
+	 * @param mixed $result Result of any other authentication errors.
+	 *
+	 * @return bool|null|WP_Error
+	 */
+	public function authenticate( $result ) {
+
+		// Another authentication method was used.
+		if ( ! is_null( $result ) ) {
+			return $result;
+		}
+
+		/**
+		 * Check for REST request.
+		 *
+		 * @param bool $rest_request Whether or not this is a REST request.
+		 */
+		$rest_request = apply_filters( 'rest_authentication_is_rest_request', ( defined( 'REST_REQUEST' ) && REST_REQUEST ) );
+
+		// This is not the authentication you're looking for.
+		if ( ! $rest_request ) {
+			return $result;
+		}
+
+		// Authentication is not required.
+		if ( ! $this->require_token() ) {
+			return $result;
+		}
+
+		// Validate the bearer token.
+		$token = $this->validate_token();
+		if ( is_wp_error( $token ) ) {
+			return $token;
+		}
+
+		// If it's a wp_user based token, set the current user.
+		if ( 'wp_user' === $token->data->user->type ) {
+			wp_set_current_user( $token->data->user->id );
+		}
+
+		// Authentication succeeded.
+		return true;
+	}
+
+	/**
+	 * Determine if the request needs to be JWT authenticated.
+	 *
+	 * @since 0.1
+	 *
+	 * @return bool
+	 */
+	public function require_token() {
+		$require_token  = true;
+		$request_uri    = isset( $_SERVER['REQUEST_URI'] ) ? $_SERVER['REQUEST_URI'] : false; // phpcs:ignore
+		$request_method = isset( $_SERVER['REQUEST_METHOD'] ) ? $_SERVER['REQUEST_METHOD'] : false; // phpcs:ignore
```

Comment:
Created https://core.trac.wordpress.org/ticket/46907 so we don't have to rely on `$_SERVER`.